src/internal/alchemy-websocket-provider.ts

```diff
@@ -193,9 +341,363 @@ export class AlchemyWebSocketProvider
   isCommunityResource(): boolean {
     return this.apiKey === DEFAULT_ALCHEMY_API_KEY;
   }
+
+  private addSocketListeners(): void {
+    this._websocket.addEventListener('message', this.handleMessage.bind(this));
+    this._websocket.addEventListener('reopen', this.handleReopen.bind(this));
+    this._websocket.addEventListener(
+      'down',
+      this.stopHeartbeatAndBackfill.bind(this)
+    );
+  }
+
+  private removeSocketListeners(): void {
+    this._websocket.removeEventListener('message', this.handleMessage);
+    this._websocket.removeEventListener('reopen', this.handleReopen);
+    this._websocket.removeEventListener('down', this.stopHeartbeatAndBackfill);
+  }
```

Comment:
This removal code won't work: you need to pass the same callback instance to `removeEventListener` that was originally passed to `addEventListener`, but here the callback added is a different instance because `.bind` returns a new instance.

One way to avoid this is to leave out the `.bind` and make sure that all three of `handleMessage`, `handleReopen`, and `stopHeartbeatAndBackfill` are bound methods. That is, they should all be written in the style:
```ts
private handleMessage = (event: MessageEvent): void => {
  // ...
}
```
and not in the style
```ts
private handleMessage(event: MessageEvent): void {
  // ...
}
```
At the moment, it appears that `handleMessage` is bound but the other two are not.

If you find that syntax too finicky, you can also store `this.handleMessage.bind(this)`, etc. in fields so you can reference the same instance in both add and remove.